src/storage/src/sink/iceberg.rs

```diff
@@ -0,0 +1,1487 @@
+    let table_ident = TableIdent::new(namespace_ident.clone(), table_name.clone());
+
+    // Try to load the table first
+    match catalog.load_table(&table_ident).await {
+        Ok(table) => {
+            // Table exists, return it
+            // TODO: Add proper schema evolution/validation to ensure compatibility
+            Ok(table)
+        }
+        Err(err) => {
+            if matches!(err.kind(), ErrorKind::TableNotFound { .. }) {
+                // Table doesn't exist, create it
+                // Note: location is not specified, letting the catalog determine the default location
+                // based on its warehouse configuration
+                let table_creation = TableCreation::builder()
+                    .name(table_name.clone())
+                    .schema(schema.clone())
+                    // Use unpartitioned spec by default
+                    // TODO: Consider making partition spec configurable
+                    // .partition_spec(UnboundPartitionSpec::builder().build())
+                    .build();
+
+                catalog
+                    .create_table(&namespace_ident, table_creation)
+                    .await
+                    .with_context(|| {
+                        format!(
+                            "Failed to create Iceberg table '{}' in namespace '{}'",
+                            table_name, namespace
+                        )
+                    })
+            } else {
+                // Some other error occurred
+                Err(err).context("Failed to load Iceberg table")
+            }
+        }
+    }
+}
+
+/// Find the most recent Materialize frontier from Iceberg snapshots.
+/// We store the frontier in snapshot metadata to track where we left off after restarts.
+/// Snapshots with operation="replace" (compactions) don't have our metadata and are skipped.
+fn retrieve_upper_from_snapshots(
+    snapshots: impl IntoIterator<Item = Arc<Snapshot>>,
+) -> anyhow::Result<Option<(Antichain<Timestamp>, Option<u64>)>> {
+    let snapshots = snapshots
+        .into_iter()
+        .sorted_by(|a, b| Ord::cmp(&b.sequence_number(), &a.sequence_number()));
+
+    for snapshot in snapshots {
+        if snapshot
+            .summary()
+            .additional_properties
+            .contains_key("mz-frontier")
+        {
+            let frontier_json = snapshot
+                .summary()
+                .additional_properties
+                .get("mz-frontier")
+                .unwrap();
+            let frontier: Vec<Timestamp> = serde_json::from_str(frontier_json)
+                .expect("Failed to deserialize frontier from snapshot properties");
+            let frontier = Antichain::from_iter(frontier);
+
+            let sink_version = snapshot
+                .summary()
+                .additional_properties
+                .get("mz-sink-version")
+                .and_then(|s| s.parse::<u64>().ok());
```

Comment:
If this parse errors we need to bubble it up, the `.ok()` here will silently ignore it and not report a version